cmd/erasure-server-pool-decom.go

```diff
@@ -0,0 +1,632 @@
+	// installation with pool removed)
+	meta.Version = "1"
+	for idx := range z.serverPools {
+		meta.Pools = append(meta.Pools, PoolInfo{
+			ID:         idx,
+			Suspend:    false,
+			LastUpdate: time.Now(),
+		})
+	}
+	if err = meta.save(ctx, z.serverPools[0]); err != nil {
+		return err
+	}
+	z.poolMeta = meta
+	return nil
+}
+
+func (z *erasureServerPools) drainObject(ctx context.Context, bucket string, gr *GetObjectReader) (err error) {
+	defer gr.Close()
+	objInfo := gr.ObjInfo
+	if objInfo.Multipart {
+		uploadID, err := z.NewMultipartUpload(ctx, bucket, objInfo.Name,
+			ObjectOptions{
+				VersionID:   objInfo.VersionID,
+				MTime:       objInfo.ModTime,
+				UserDefined: objInfo.UserDefined,
+			})
+		if err != nil {
+			return err
+		}
+		defer z.AbortMultipartUpload(ctx, bucket, objInfo.Name, uploadID, ObjectOptions{})
+		var parts = make([]CompletePart, 0, len(objInfo.Parts))
+		for _, part := range objInfo.Parts {
+			hr, err := hash.NewReader(gr, part.Size, "", "", part.Size)
+			if err != nil {
+				return err
+			}
+			_, err = z.PutObjectPart(ctx, bucket, objInfo.Name, uploadID,
+				part.Number,
+				NewPutObjReader(hr),
+				ObjectOptions{})
+			if err != nil {
+				return err
+			}
+			parts = append(parts, CompletePart{
+				PartNumber: part.Number,
+				ETag:       part.ETag,
+			})
+		}
+		_, err = z.CompleteMultipartUpload(ctx, bucket, objInfo.Name, uploadID, parts, ObjectOptions{
+			MTime: objInfo.ModTime,
+		})
+		return err
+	}
+	hr, err := hash.NewReader(gr, objInfo.Size, "", "", objInfo.Size)
+	if err != nil {
+		return err
+	}
+	_, err = z.PutObject(ctx,
+		bucket,
+		objInfo.Name,
+		NewPutObjReader(hr),
+		ObjectOptions{
+			VersionID:   objInfo.VersionID,
+			MTime:       objInfo.ModTime,
+			UserDefined: objInfo.UserDefined,
+		})
+	return err
+}
+
+func (z *erasureServerPools) drainInBackground(idx int, buckets []BucketInfo) error {
+	pool := z.serverPools[idx]
+	for _, bi := range buckets {
+		versioned := globalBucketVersioningSys.Enabled(bi.Name)
+		for _, set := range pool.sets {
+			disks := set.getOnlineDisks()
+			if len(disks) == 0 {
+				logger.LogIf(GlobalContext, fmt.Errorf("no online disks found for set with endpoints %s",
+					set.getEndpoints()))
+				continue
+			}
+
+			drainEntry := func(entry metaCacheEntry) {
+				if entry.isDir() {
+					return
+				}
+
+				fivs, err := entry.fileInfoVersions(bi.Name)
+				if err != nil {
+					return
+				}
+
+				// we need a reversed order for Draining,
+				// to create the appropriate stack.
+				versionsSorter(fivs.Versions).reverse()
+
+				for _, version := range fivs.Versions {
+					// Skip transitioned objects for now.
+					if version.IsRemote() {
+						continue
+					}
+					// We will skip draining delete markers
+					// with single version, its as good as
+					// there is no data associated with the
+					// object.
+					if version.Deleted && len(fivs.Versions) == 1 {
+						continue
+					}
+					if version.Deleted {
```

Comment:
Maybe let's move this code to drainObject as well, it is good to have a low level draining object function in one function